charms/knative-serving/src/charm_template.py

```diff
@@ -0,0 +1,248 @@
+            status = ActiveStatus()
+        else:
+            # Hit one or more errors with resources.  Return status for worst and log all
+            self.log.info("Charm is not active due to one or more issues:")
+
+            # Log all errors, ignoring None's
+            errors = [error for error in errors if error is not None]
+            for i, error in enumerate(errors):
+                self.log.info(f"Issue {i+1}/{len(errors)}: {error.msg}")
+
+            # Return status based on the worst thing we encountered
+            status = get_first_worst_error(errors).status
+
+        return status
+
+    def _check_leader(self):
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            raise LeadershipError("Waiting for leadership", WaitingStatus)
+
+    def on_config_changed(self, event):
+        self.on_install(event)
+
+    def on_install(self, event):
+        """Installs all objects, checking and setting status at the end of install"""
+        self.unit.status = MaintenanceStatus("Deploying resources to cluster")
+        try:
+            self._check_leader()
+            resources = self.render_manifests()
+            self.reconcile(resources)
+        except ErrorWithStatus as e:
+            self.log_and_set_status(e.status)
+            return
+
+        self.log_and_set_status(self.charm_status(resources))
+
+    def on_leader_elected(self, event):
+        self.on_install(event)
+
+    def on_remove(self, _):
+        """Remove all objects from the cluster
+
+        By default, this should be implicitly handled by Juju and specific resource removal should
+        not be necessary, but in some situations that is not the case.  Override this method with
+        custom removal logic if required.
+        """
+        self.log.info("Removal hook fired but no custom removal actions taken - exiting on_remove")
+        pass
+
+    def on_upgrade_charm(self, event):
+        self.on_install(event)
+
+    def on_update_status(self, event):
+        charm_status = self.charm_status()
+        if not isinstance(charm_status, ActiveStatus):
+            # Queue an install event to try to reconcile our resources
+            self.log.info("Charm emitting an install event to try to reconcile resources")
+        self.on.install.emit()
+        self.log_and_set_status(charm_status)
+
+    def reconcile(self, resources: Optional[List[Union[NamespacedResource, GlobalResource]]] = None):
+        """Reconcile our Kubernetes objects to achieve the desired state
+
+        This can be invoked to both install or update objects in the cluster.  It uses an apply
+        logic to update things only if necessary.  This method by default __does not__ remove
+        objects that are no longer required - if handling that situation is required, it must be
+        done separately.
+        """
+        self.log.info("Reconciling")
```

Comment:
Should `reconcile` have a leadership check?  I think it should probably exit without doing anything if not the leader, but I'm not certain...